src/ethereum/frontier/vm/runtime.py

```diff
@@ -0,0 +1,60 @@
+"""
+Ethereum Virtual Machine (EVM) Runtime Operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Runtime related operations used while executing EVM code.
+"""
+from typing import Set
+
+from ethereum.base_types import Uint
+
+from .instructions import Ops
+
+
+def get_valid_jump_destinations(code: bytes) -> Set[Uint]:
+    """
+    Analyze the evm code to obtain the set of valid jump destinations.
+
+    Valid jump destinations are defined as follows:
+        * The jump destination is less than the length of the code.
+        * The jump destination should have the `JUMPDEST` opcode (0x5B).
+        * The jump destination shouldn't be part of the data corresponding to
+          `PUSH-N` opcodes.
+
+    Note - Jump destinations are 0-indexed.
+
+    Parameters
+    ----------
+    code :
+        The EVM code which is to be executed.
+
+    Returns
+    -------
+    valid_jump_destinations: `Set[Uint]`
+        The set of valid jump destinations in the code.
+    """
+    valid_jump_destinations = set()
+    pc = Uint(0)
+
+    while pc < len(code):
+        current_opcode = Ops(code[pc])
```

Comment:
I think we need to handle `ValueError` and treat it as a a non-jumpdest non-push opcode?

Reply:
I raised `InvalidOpcode` here instead of `InvalidJumpDestError` as the first error seemed apt here. Let me know if you want it the other way around. Thanks.

Reply:
We can't raise here because the opcode wasn't actually executed. For example:

```
push1 0x04
jump
invalid
jumpdest
...
```

The invalid opcode is never executed, and doesn't mess up the jumpdest analysis.

Reply:
Ok, I'll just skip invalid opcodes while performing jumpdest analysis, without raising any error.